File: backtracking/n_queens.cpp
#include <functional>
#include <iostream>
#include <string>
#include <vector>

std::string FormatBoard(uint64_t n, const std::vector<std::string>& board)
{
    std::string formatted(2 * n * n, 0);
    size_t i = -1;
    for (uint64_t row = n - 1; row != -1; --row)
    {
        for (char ch : board[row])
        {
            formatted[++i] = ch;
            formatted[++i] = ' ';
        }
        formatted[i - 1] = '\n';
    }
    return formatted;
}

constexpr uint64_t Diag(uint64_t n, uint64_t row, uint64_t col)
{
    return n + row - col - 1;
}

constexpr uint64_t Anti(uint64_t n, uint64_t row, uint64_t col)
{
    return row + col;
}

void n_queens(int n)
{
    std::vector<bool> cols(n, true);
    std::vector<bool> diags(2 * n - 1, true);
    std::vector<bool> antis(2 * n - 1, true);
    std::vector<std::string> board(n, std::string(n, '.'));

    std::function<bool(uint64_t, uint64_t)> can_put = [&](uint64_t row, uint64_t col)
    {
        return cols[col] && diags[Diag(n, row, col)] && antis[Anti(n, row, col)];
    };
    std::function<void(uint64_t, uint64_t)> put_queen = [&](uint64_t row, uint64_t col)
    {
        cols[col] = false;
        diags[Diag(n, row, col)] = false;
        antis[Anti(n, row, col)] = false;
        board[row][col] = 'Q';
    };
    std::function<void(uint64_t, uint64_t)> remove_queen = [&](uint64_t row, uint64_t col)
    {
        cols[col] = true;
        diags[Diag(n, row, col)] = true;
        antis[Anti(n, row, col)] = true;
        board[row][col] = '.';
    };
    std::function<void(uint64_t, uint64_t)> worker = [&](uint64_t no_queens, uint64_t row)
    {
        if (no_queens)
        {
            for (uint64_t col = 0; col < n; ++col)
            {
                if (can_put(row, col))
                {
                    put_queen(row, col);
                    worker(no_queens - 1, row + 1);
                    remove_queen(row, col);
                }
            }
        }
        else
        {
            std::cout << FormatBoard(n, board) << std::endl;
        }
    };
    worker(n, 0);
}

int main(int argc, char* argv[])
{
    for (size_t i = 1; i < argc; ++i)
    {
        n_queens(std::stoi(argv[i]));
    }
}

File: backtracking/n_queens.py
import copy
import itertools
import sys

from typing import List

def n_queens(n: int) -> int:

    def make_board(n: int) -> List[List[str]]:
        return [['.'] * n for i in range(n)]
    
    def format_board(board: List[List[str]]) -> str:
        return '\n'.join(' '.join(row) for row in reversed(board))
        
    def diag(row: int, col: int) -> int:
        return n + row - col - 1
    
    def anti(row: int, col: int) -> int:
        return row + col

    def can_put(row: int, col: int) -> bool:
        return cols[col] and diags[diag(row, col)] and antis[anti(row, col)]

    def put_queen(row: int, col: int) -> None:
        cols[col] = False
        diags[diag(row, col)] = False
        antis[anti(row, col)] = False
        board[row][col] = 'Q'
        
    def remove_queen(row: int, col: int) -> None:
        cols[col] = True
        diags[diag(row, col)] = True
        antis[anti(row, col)] = True
        board[row][col] = '.'
    
    def worker(no_queens: int, row: int) -> None:
        if no_queens:
            for col in range(n):
                if can_put(row, col):
                    put_queen(row, col)
                    worker(no_queens - 1, row + 1)
                    remove_queen(row, col)
        else:
            print('Solution #%i:\n%s' % (next(sol_count) + 1, format_board(board)))

    cols = [True] * n
    diags = [True] * (2 * n - 1)
    antis = [True] * (2 * n - 1)
    board = make_board(n)
    sol_count = itertools.count()
    worker(n, 0)
    return next(sol_count)

if __name__ == '__main__':
    for arg in sys.argv:
        if arg.isdigit():
            print('n_queens(%s) = %i' % (arg, n_queens(int(arg))))